mod.go

```diff
@@ -0,0 +1,624 @@
+	return c.hit, c.miss
+}
+
+var (
+	// Fixed modulus and its reciprocal
+	fixed_m Int
+	fixed_r [5]uint64
+)
+
+func init() {
+	if fixedModulus {
+		// Initialise fixed modulus
+		fixed_m[3] = 0xffffffff00000001
+		fixed_m[2] = 0x0000000000000000
+		fixed_m[1] = 0x00000000ffffffff
+		fixed_m[0] = 0xffffffffffffffff
+
+		fixed_r = reciprocal(fixed_m)
+	}
+}
+
+func (cache *reciprocalCache) has(m Int, index uint64, dest *[5]uint64) bool {
+	if fixedModulus && m.Eq(&fixed_m) {
+		dest[0] = fixed_r[0]
+		dest[1] = fixed_r[1]
+		dest[2] = fixed_r[2]
+		dest[3] = fixed_r[3]
+		dest[4] = fixed_r[4]
+
+		return true
+	}
+
+	if cacheWays == 0 {
+		return false
+	}
+
+	cache.set[index].rw.RLock()
+	defer cache.set[index].rw.RUnlock()
+
+	for w := 0; w < cacheWays; w++ {
+		if cache.set[index].mod[w].Eq(&m) {
+			copy(dest[:], cache.set[index].inv[w][:])
+			cache.hit++
+			return true
+		}
+	}
+	cache.miss++
+	return false
+}
+
+func (cache *reciprocalCache) put(m Int, index uint64, mu [5]uint64) {
+	if cacheWays == 0 {
+		return
+	}
+
+	cache.set[index].rw.Lock()
+	defer cache.set[index].rw.Unlock()
+
+	var w int
+	for w = 0; w < cacheWays; w++ {
+		if cache.set[index].mod[w].IsZero() {
+			// Found an empty slot
+			cache.set[index].mod[w] = m
+			cache.set[index].inv[w] = mu
+			return
+		}
+	}
+	// Shift old elements, evicting the oldest
+	for w = cacheWays - 1; w > 0; w-- {
+		cache.set[index].mod[w] = cache.set[index].mod[w-1]
+		cache.set[index].inv[w] = cache.set[index].inv[w-1]
+	}
+	// w == 0
+	cache.set[index].mod[w] = m
+	cache.set[index].inv[w] = mu
+}
+
+var cache reciprocalCache
+
+// Some utility functions
+
+func leadingZeros(x Int) (z int) {
+	z  = bits.LeadingZeros64(x[3]); if z <  64 { return z }
+	z += bits.LeadingZeros64(x[2]); if z < 128 { return z }
+	z += bits.LeadingZeros64(x[1]); if z < 192 { return z }
+	z += bits.LeadingZeros64(x[0]); return z
+}
+
+func onesCount(x Int) (z int) {
+	z =	bits.OnesCount64(x[0]) +
+		bits.OnesCount64(x[1]) +
+		bits.OnesCount64(x[2]) +
+		bits.OnesCount64(x[3])
+	return z
+}
+
+// shiftright320 shifts a 320-bit value 0-63 bits right
+// z = x >> (s % 64)
+
+func shiftright320(x [5]uint64, s uint) (z [5]uint64) {
+	r := s % 64	// right shift
+	l := 64 - r	// left shift
+
+	z[0] = (x[0] >> r) | (x[1] << l)
+	z[1] = (x[1] >> r) | (x[2] << l)
+	z[2] = (x[2] >> r) | (x[3] << l)
+	z[3] = (x[3] >> r) | (x[4] << l)
+	z[4] = (x[4] >> r)
+
+	return z
+}
+
+// reciprocal computes a 320-bit value representing 1/m
+//
+// Notes:
+// - reciprocal(0) = 0
+// - reciprocal(1) = 2^320-1
+// - otherwise, the result is normalised to have non-zero most significant word
+// - starts with a 32-bit division, refines with newton-raphson iterations
+
+func reciprocal(m Int) (mu [5]uint64) {
+
+	s := leadingZeros(m)
+	p := 255 - s // floor(log_2(m)), m>0
+
+	// 0 or a power of 2?
+
+	if onesCount(m) <= 1 {
+		if s >= 255 { // m <= 1
+			mu[4] = -m[0]
+			mu[3] = -m[0]
+			mu[2] = -m[0]
+			mu[1] = -m[0]
+			mu[0] = -m[0]
```

Comment:
Unreachable via mulmod, which only calls `reciprocal` if `m[3]!=0`, so `leadingZeroes(m)` can never be `>63`. 

Either we need specific tests on `reciprocal` to cover this, or this can be removed, at a loss of generality.

Reply:
I tried to add this: 
```golang
	t.Run("CustomMulMod", func(t *testing.T) {
		// CustomMulMod is a special case. Normally, we don't use the
		// reciprocal-approach if the modulus is < 192 bits. This
		// uses the reciprocal-approach in all cases, for better
		// coverage

		mulmodCustom := func(z, x, y, m *Int) *Int {
			if x.IsZero() || y.IsZero() || m.IsZero() {
				return z.Clear()
			}
			p := umul(x, y)
			mu := reciprocal(*m)
			r := reduce4(p, *m, mu)
			return z.Set(&r)
		}
		proc(t, mulmodCustom, func(z, x, y, m *big.Int) *big.Int {
			if m.Sign() == 0 {
				return z.SetUint64(0)
			}
			return mulMod(z, x, y, m)
		})
	})
```
But that causes the test to hang. I'm not sufficiently good at math to figure out how to do it correctly